### ISSUE TITLE
Couldn't `iretq` to userspace in SMP EFI boot

### ISSUE THREAD

**Issue by junyang-zh** [asterinas/asterinas#1721]
### Describe the bug

When running with Linux EFI handover 64 protocol and SMP, the kernel fails on the first entry to user space. QEMU aborts immediately on the `iretq` instruction, without any other hints of what's going on.

On single threaded or multiboot, it works just fine. Only SMP & EFI boot is affected.

### To Reproduce

```bash
make run SMP=2 BOOT_PROTOCOL=linux-efi-handover64
```

Release or debug profile does not matter. With KVM on, the kernel must fail on first `iretq`. With KVM off, the kernel may fail on the first or second `sysretq` afterwards.

**Comment by junyang-zh** [comment]
I suspect that there's some difference of SMP GDT setup when working with EFI boot. @cqs21 have a clue?

**Comment by cqs21** [comment]
> I suspect that there's some difference of SMP GDT setup when working with EFI boot. @cqs21 have a clue?

@junyang-zh You're right. The `GDT` for the bsp and the ap are not identical. There is a quick fix, but the root cause seems to be that `USER_SS` and `USER_CS` are not `cpu_local`.

```
diff --git a/ostd/src/arch/x86/trap/gdt.rs b/ostd/src/arch/x86/trap/gdt.rs
index 9511dd48..1b5277da 100644
--- a/ostd/src/arch/x86/trap/gdt.rs
+++ b/ostd/src/arch/x86/trap/gdt.rs
@@ -61,7 +61,9 @@ pub unsafe fn init(on_bsp: bool) {
     //   STAR[47:32] = K_CS   = K_SS - 8
     //   STAR[63:48] = U_CS32 = U_SS32 - 8 = U_CS - 16
     let mut gdt = Vec::from(old_gdt);
+    gdt.resize(20, 0);
     gdt.extend([tss0, tss1, KCODE64, KDATA64, UCODE32, UDATA32, UCODE64].iter());
+
     let gdt = Vec::leak(gdt);

     // Load new GDT and TSS.
@@ -70,12 +72,12 @@ pub unsafe fn init(on_bsp: bool) {
         base: VirtAddr::new(gdt.as_ptr() as _),
     });
     load_tss(SegmentSelector::new(
-        entry_count as u16,
+        20 as u16,
         PrivilegeLevel::Ring0,
     ));

-    let sysret = SegmentSelector::new(entry_count as u16 + 4, PrivilegeLevel::Ring3).0;
-    let syscall = SegmentSelector::new(entry_count as u16 + 2, PrivilegeLevel::Ring0).0;
+    let sysret = SegmentSelector::new(20 as u16 + 4, PrivilegeLevel::Ring3).0;
+    let syscall = SegmentSelector::new(20 as u16 + 2, PrivilegeLevel::Ring0).0;
     Star::write_raw(sysret, syscall);

     USER_SS = sysret + 8;
```

**Comment by Hsy-Intel** [comment]
Hi @cqs21, does this quick fix works for aps' idt init? Seems that the aps can't handle virtualization exception in trap handler correctly in my Intel TDX environment.

**Comment by Hsy-Intel** [comment]
Hi @cqs21, does this quick fix works for aps' idt init? Seems that the aps can't handle virtualization exception in kernel trap handler correctly in my Intel TDX env.

**Comment by cqs21** [comment]
> does this quick fix works for aps' idt init?

@Hsy-Intel It may not work, since `USER_SS` and `USER_CS` are not utilized when returning from the kernel trap handler. I'm not familiar with `#VE`, but there might be another error.

**Comment by Hsy-Intel** [comment]
> > does this quick fix works for aps' idt init?
> 
> @Hsy-Intel It may not work, since `USER_SS` and `USER_CS` are not utilized when returning from the kernel trap handler. I'm not familiar with `#VE`, but there might be another error.

I'm not sure if the APs in kernel mode can correctly enter the trap handler and handle page faults. If they can, it indicates an issue with the handling logic of VE in TDX; if they can't, it suggests an unknown issue in the protocol.

**Comment by lrh2000** [comment]
This has been fixed in https://github.com/asterinas/asterinas/commit/dd3aa8fe81d6ace633c4e6bc82b1390d2808d14c by https://github.com/asterinas/asterinas/pull/1651.
